### PR TITLE
Align model with template

### PR DESCRIPTION
**Pull request by ClemensLinnhoff**
**Add a description**
This PR aligns the camera model with the [template](https://github.com/openMSL/sl-1-0-sensor-model-repository-template) inkl. documentation structure and CI test pipeline.

**Take this checklist as orientation for yourself, if this PR is ready for Maintainer Review**
- [x] My suggestion follows the [governance rules](https://github.com/openMSL/governance-and-documentation).
- [x] All commits of this PR are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).
- [x] My changes generate no errors when passing CI tests. 
- [x] I updated all documentation (readmes incl. figures) according to my changes.
- [x] I have successfully implemented and tested my fix/feature locally.
- [x] Appropriate reviewer(s) are assigned.

### CODE REVIEW

File: doc/README.md
Comment (ClemensLinnhoff) [inline doc/README.md:1]: Please put the contents of the readme in the readme template in the main folder.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+set(LINK_WITH_SHARED_OSI OFF CACHE BOOL "Link FMU with shared OSI library instead of statically linking")
+set(PUBLIC_LOGGING OFF CACHE BOOL "Enable logging via FMI logger")
+set(PRIVATE_LOGGING OFF CACHE BOOL "Enable private logging to file")
+
+string(TIMESTAMP FMUTIMESTAMP UTC)
+string(MD5 FMUGUID modelDescription.in.xml)
+configure_file(modelDescription.in.xml modelDescription.xml @ONLY)
+
+find_package(Protobuf 2.6.1 REQUIRED)
+add_library(COSMPCameraSensor SHARED COSMPCameraSensor.cpp COSMPCameraSensor.h)
```
Comment (ClemensLinnhoff) [inline src/CMakeLists.txt]: In this line it has to say "OSMPCameraSensor.cpp OSMPCameraSensor.h" without the "C"s at the beginning.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	endif()
+endif()
+
+add_custom_command(TARGET COSMPCameraSensor
+		POST_BUILD
+		COMMAND ${CMAKE_COMMAND} -E remove_directory "${CMAKE_CURRENT_BINARY_DIR}/buildfmu"
+		COMMAND ${CMAKE_COMMAND} -E make_directory "${CMAKE_CURRENT_BINARY_DIR}/buildfmu/sources"
+		COMMAND ${CMAKE_COMMAND} -E make_directory "${CMAKE_CURRENT_BINARY_DIR}/buildfmu/binaries/${FMI_BINARIES_PLATFORM}"
+		COMMAND ${CMAKE_COMMAND} -E copy "${CMAKE_CURRENT_BINARY_DIR}/modelDescription.xml" "${CMAKE_CURRENT_BINARY_DIR}/buildfmu"
+		COMMAND ${CMAKE_COMMAND} -E copy "${CMAKE_CURRENT_SOURCE_DIR}/COSMPCameraSensor.cpp" "${CMAKE_CURRENT_BINARY_DIR}/buildfmu/sources/"
```
Comment (ClemensLinnhoff) [inline src/CMakeLists.txt]: Same as above with the .cpp and the .h file

**Comment by ClemensLinnhoff** [comment]
I just ran clang-format again locally and it fixed some further minor format violations. I don't know, why the pipeline still shows some violations. Formatting looks fine to me.

However, there are still some clang-tidy violations worth fixing. E.g. some variables are initialized, but never used. From the looks of it, they were intended for logging, but the logs have been commented out, e.g. in /src/OSMPCameraSensor.cpp:634 (there are also others). And maybe you can also fix the "narrowing conversion from 'uint64_t'" remarks. The other annotations are fine with me if we ignore them.

One more thing: You can delete the remark in <> at the beginning of the Readme. And you can delete the Readme file in the doc folder, since its contents is already contained in the main readme.

The main branch contains a video of the model. Would it be possible to convert the video into a gif and include it in the readme? Seeing the model in action is always a good eye-catcher.

**Comment by ClemensLinnhoff** [comment]
And you need to sign-off your commits. Currently they are verified, so the GPG-sign works, but the DCO sign-off is missing.

**Comment by MarzenaFranek** [comment]
Habe ich versucht. Wie erkennt man ob es funktioniert hat?

Mit freundlichen Grüßen / Best regards

Dr. Marzena Franek

(CR/AEL2)
Robert Bosch GmbH | Renningen | 70465 Stuttgart | GERMANY | www.bosch.com<http://www.bosch.com/>
Tel. +49 711 811-17691 | Mobil +49 1522 9403572 | Telefax +49 711 811 | ***@***.******@***.***>

Sitz: Stuttgart, Registergericht: Amtsgericht Stuttgart, HRB 14000;
Aufsichtsratsvorsitzender: Prof. Dr. Stefan Asenkerschbaumer; Geschäftsführung: Dr. Stefan Hartung,
Dr. Christian Fischer, Dr. Markus Forschner, Stefan Grosch, Dr. Markus Heyn, Dr. Tanja Rückert
​
Von: Clemens Linnhoff ***@***.***>
Gesendet: Mittwoch, 12. April 2023 17:16
An: openMSL/sl-1-4-object-based-camera-object-model ***@***.***>
Cc: Franek Marzena (CR/AEL2) ***@***.***>; Assign ***@***.***>
Betreff: Re: [openMSL/sl-1-4-object-based-camera-object-model] Align model with template (PR #1)


And you need to sign-off you commits. Currently they are verified, so the GPG-sign works, but the DCO sign-off is missing.

**Comment by ClemensLinnhoff** [comment]
Hi, bei einem Commit hat es funktioniert:
[cid:ff6a2ee7-7fa3-413e-9e74-f5383b50c0a3]

Der letzte Commit war aber nicht mehr signed-off:
[cid:e357f525-eede-41ed-8f2c-913b36349ff1]

Wenn der signed-off ist, kannst du neben der Commit Message auf die drei Punkte gehen und siehst das dann.

Schöne Grüße
Clemens

__________________________________
Dr.-Ing. Clemens Linnhoff
CTO

Persival GmbH
Carlo-Mierendorff-Str. 2
64372 Ober-Ramstadt
Tel.: +49 6154 803659-3
www.persival.de<https://www.persival.de/>



________________________________

**Comment by ClemensLinnhoff** [comment]
Corresponding to your email @MarzenaFranek, I fixed the integer conversion remark from clang-tidy. I think this concludes working on the clang-tidy remarks. The remaining ones are fine by me.
While I was fixing that, I found several unused parameters in the Lamda functions, e.g. &mpos_x, &mpos_y, &mpos_z, &origin_rot_x, &origin_rot_y, &origin_rot_z, &mpos_rot_x, &mpos_rot_y, &mpos_rot_z, and actual_range in the for_each loop in line 1087 and similarly in line 1307, line 1448 and line 1607.
Could you fix those? Then I think the C++ code would be finished.

I also removed the image from the readme, since you added the video as a gif.
While I was doing that, I noticed the following section in the modeling approach:
> The actual logic of the model is packed in a so called strategy.
The apply function of the strategy is called by the do_calc function of the OSMPFramework.
The strategy itself is structured into four modules as shown in the image below.

I think this is a copy-paste-error, since strategies are only used in the FZD OSMP Framework, which is not used in this model. I could also not find the modules described in the block diagram, such as clustering, in the code. Also some of the interfaces seem odd. To my knowledge there is no "osi3::VideoDetection" in OSI.
Could you go over the readme and check these sections?

**Comment by MarzenaFranek** [comment]
Done !

Mit freundlichen Grüßen / Best regards

Dr. Marzena Franek

(CR/AEL2)
Robert Bosch GmbH | Renningen | 70465 Stuttgart | GERMANY | www.bosch.com<http://www.bosch.com/>
Tel. +49 711 811-17691 | Mobil +49 1522 9403572 | Telefax +49 711 811 | ***@***.******@***.***>

Sitz: Stuttgart, Registergericht: Amtsgericht Stuttgart, HRB 14000;
Aufsichtsratsvorsitzender: Prof. Dr. Stefan Asenkerschbaumer; Geschäftsführung: Dr. Stefan Hartung,
Dr. Christian Fischer, Dr. Markus Forschner, Stefan Grosch, Dr. Markus Heyn, Dr. Tanja Rückert
​
Von: Clemens Linnhoff ***@***.***>
Gesendet: Freitag, 14. April 2023 13:38
An: openMSL/sl-1-4-object-based-camera-object-model ***@***.***>
Cc: Franek Marzena (CR/AEL2) ***@***.***>; Mention ***@***.***>
Betreff: Re: [openMSL/sl-1-4-object-based-camera-object-model] Align model with template (PR #1)


Corresponding to your email @MarzenaFranek<https://github.com/MarzenaFranek>, I fixed the integer conversion remark from clang-tidy. I think this concludes working on the clang-tidy remarks. The remaining ones are fine by me.
While I was fixing that, I found several unused parameters in the Lamda functions, e.g. &mpos_x, &mpos_y, &mpos_z, &origin_rot_x, &origin_rot_y, &origin_rot_z, &mpos_rot_x, &mpos_rot_y, &mpos_rot_z, and actual_range in the for_each loop in line 1087 and similarly in line 1307, line 1448 and line 1607.
Could you fix those? Then I think the C++ code would be finished.

I also removed the image from the readme, since you added the video as a gif.
While I was doing that, I noticed the following section in the modeling approach:

The actual logic of the model is packed in a so called strategy.
The apply function of the strategy is called by the do_calc function of the OSMPFramework.
The strategy itself is structured into four modules as shown in the image below.

I think this is a copy-paste-error, since strategies are only used in the FZD OSMP Framework, which is not used in this model. I could also not find the modules described in the block diagram, such as clustering, in the code. Also some of the interfaces seem odd. To my knowledge there is no "osi3::VideoDetection" in OSI.
Could you go over the readme and check these sections?

**Comment by ClemensLinnhoff** [comment]
Looks very good! I tried to fix the last clang-format remarks, but in my opinion the formatting is fine.

The DCO sign-off check fails, as not all commit are signed off. But since your last ones are, and we are doing a squash merge anyways, I am fine with ignoring it. Therefore, I give my approval to the PR.
Before merging, we need one more review from another maintainer.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
+name: Build
+
+on:
+  workflow_call:
+
+env:
+  FMU_FILE_NAME: OSMPCameraSensor.fmu #TODO: Adjust to file name of your FMU
```
Comment (jdsika) [inline .github/workflows/build.yml:7]: @ClemensLinnhoff this is outdated now, as we use the repo name as FMU name, right?

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,24 +1,18 @@
...
-set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/Modules/")
+cmake_policy(SET CMP0048 NEW)
+cmake_minimum_required(VERSION 3.12)
 
-# Set a default build type if none was specified
-set(default_build_type "Release")
-if(EXISTS "${CMAKE_SOURCE_DIR}/../.git")
-  set(default_build_type "Debug")
-endif()
+project(OSMPCameraSensor)
```
Comment (jdsika) [inline CMakeLists.txt:4]: @ClemensLinnhoff would this also need to be adapted to the repo name?

File: OSMPCameraSensor/.depend
Comment (jdsika) [inline OSMPCameraSensor/.depend:1]: Is this empty file needed?

**Comment by jdsika** [comment]
For the sake of showing the use of the MPL 2.0 OSMP sensor packaging in combination with a model implementation and looking at the weak copy left of MPL, shouldn't we split the function into a different file? In this case the function will also be MPL but we had the diuscussion about how this could be used in combination with other licenses? @ClemensLinnhoff

File: .github/workflows/build.yml (hunk 3 above)
Comment (ClemensLinnhoff) [reply .github/workflows/build.yml:7]: Yes, but let's merge this PR first, then adapt to naming the FMU according to the repo. This requirement did not exist, when @MarzenaFranek startet, so I would like to keep it separate.

File: CMakeLists.txt (hunk 4 above)
Comment (ClemensLinnhoff) [reply CMakeLists.txt:4]: Not necessarily, but as above, I would adapt this in a separate PR.

**Comment by ClemensLinnhoff** [comment]
> For the sake of showing the use of the MPL 2.0 OSMP sensor packaging in combination with a model implementation and looking at the weak copy left of MPL, shouldn't we split the function into a different file? In this case the function will also be MPL but we had the diuscussion about how this could be used in combination with other licenses? @ClemensLinnhoff

Yes, I think this would generally be a good idea! Would it be possible for you @MarzenaFranek to do this? So put the model contents (pretty much everything in the doCalc function) into a separate file? Then we would have a clean separation between model and OSMP.

**Comment by MarzenaFranek** [comment]
I can try this in KW19, currently I am full with other tasks,.

Mit freundlichen Grüßen / Best regards

Dr. Marzena Franek

(CR/AEL2)
Robert Bosch GmbH | Renningen | 70465 Stuttgart | GERMANY | www.bosch.com<http://www.bosch.com/>
Tel. +49 711 811-17691 | Mobil +49 1522 9403572 | Telefax +49 711 811 | ***@***.******@***.***>

Sitz: Stuttgart, Registergericht: Amtsgericht Stuttgart, HRB 14000;
Aufsichtsratsvorsitzender: Prof. Dr. Stefan Asenkerschbaumer; Geschäftsführung: Dr. Stefan Hartung,
Dr. Christian Fischer, Dr. Markus Forschner, Stefan Grosch, Dr. Markus Heyn, Dr. Tanja Rückert
​
Von: Clemens Linnhoff ***@***.***>
Gesendet: Montag, 17. April 2023 17:43
An: openMSL/sl-1-4-object-based-camera-object-model ***@***.***>
Cc: Franek Marzena (CR/AEL2) ***@***.***>; Mention ***@***.***>
Betreff: Re: [openMSL/sl-1-4-object-based-camera-object-model] Align model with template (PR #1)


For the sake of showing the use of the MPL 2.0 OSMP sensor packaging in combination with a model implementation and looking at the weak copy left of MPL, shouldn't we split the function into a different file? In this case the function will also be MPL but we had the diuscussion about how this could be used in combination with other licenses? @ClemensLinnhoff<https://github.com/ClemensLinnhoff>

Yes, I think this would generally be a good idea! Would it be possible for you @MarzenaFranek<https://github.com/MarzenaFranek> to do this? So put the model contents (pretty much everything in the doCalc function) into a separate file? Then we would have a clean separation between model and OSMP.

**Comment by ClemensLinnhoff** [comment]
I just added the separation to the [sl-1-0-sensor-model-repository-template](https://github.com/openMSL/sl-1-0-sensor-model-repository-template). Now there is an OSMP.cpp file with the packaging and a MySensorModel.cpp file with the actual model in the src directory.
@MarzenaFranek you can basically copy this concept to the camera model.

**Comment by ClemensLinnhoff** [comment]
Data Cluster Maintainer Decision: Merge as is, update CI pipeline according to template in new PR.

**Pull request merged**